File: DotNetGotchas/CSharp/ResourceHold/DependingOnFinalize/FinalizePeril/Wrapper.cs
//Wrapper.cs
using System;
using ACOMCompLib;

namespace FinalizePeril
{
	public class Wrapper : IDisposable
	{
		IMyComp comp = new MyCompClass();

		public int doSomething()
		{
			int result;
			comp.doSomething(out result);	
			return result;
		}

		~Wrapper()
		{
			System.Runtime.InteropServices.Marshal.ReleaseComObject(comp);
		}

		#region IDisposable Members

		public void Dispose()
		{
			System.Runtime.InteropServices.Marshal.ReleaseComObject(comp);
		}

		#endregion
	}
}

File: DotNetGotchas/VBDN/ResourceHold/DependingOnFinalize/FinalizePeril/Wrapper.vb
'Wrapper.vb

Imports ACOMCompLib

Public Class Wrapper
    Implements IDisposable
    Dim comp As IMyComp = New MyCompClass

    Public Function doSomething() As Integer
        Dim result As Integer
        comp.doSomething(result)
        Return result
    End Function

    Public Sub Dispose() Implements System.IDisposable.Dispose
        System.Runtime.InteropServices.Marshal.ReleaseComObject(comp)
    End Sub

    Protected Overrides Sub Finalize()
        MyBase.Finalize()
        System.Runtime.InteropServices.Marshal.ReleaseComObject(comp)
    End Sub
End Class
